src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java

```diff
@@ -777,7 +777,8 @@ public void evaluate() throws Throwable {
                 WorkflowRun run = createAndRunSleeperJob(story.j.jenkins, jobName, FlowDurabilityHint.MAX_SURVIVABILITY, false);
                 FlowExecution exec = run.getExecution();
                 if (exec instanceof CpsFlowExecution) {
-                    assert ((CpsFlowExecution) exec).getStorage().isPersistedFully();
+                    ((CpsFlowExecution) exec).waitForSuspension(); // till done writing head node ID into build.xml
```

Comment:
`waitForSuspension` should be deprecated. Generally there is something higher-level you want to wait for.

Reply:
> if there was discussion of deprecation

Basically dating to the early days of ~Workflow~ Pipeline when the method was even introduced. We managed to get rid of most calls and there are just a few remaining ones to be cleaned up.

> were there suggestions on what else does this better?

Depends on the circumstance. If you are waiting for a build to get a certain point, `SemaphoreStep` is the usual idiom.

I am not very familiar with the test cases here. Basically the comment

> till done writing head node ID into build.xml

should guide you. Is that the actual thing you want to wait for? If so, sleep until `build.xml` contains it.

Reply:
I do not need to wait for the build progress. I do not need the sleep step to end, we are shutting down when it's still going. We just need a little time to persist properly.

> If so, sleep until build.xml contains it.
Do you mean a cycle of checking the contents of the file and sleeping more if they are incorrect?

Reply:
Something like that, yes. (http://www.awaitility.org/ does that, but a simple `while` loop suffices.) Again, assuming that you are confident that the thing you are truly interested in waiting for is a particular field in this file.

Reply:
I honestly do not think a loop of file read is a great solution.
Firstly, now, instead of waiting _until the last head is done being reacted to and written by another thread_, we wait _until head is set to number X in the file_. That number X will be different and have to be counted separately for each pipeline if we want to reuse this pattern in some other test.
Secondly, what if a bug is introduced into the code, which the tests should be helping us find, and the correct head is _never_ written in the file? Then the waiting will simply time out, which seems less informative than restarting and getting "expected head X, got head Y" that we were getting here.

You could phrase it like that - the particular field in this file - or phrase it as waiting for the thread that reacts to a new head being set to be done with its job (writing included). `waitForSuspension` did it because the purpose of it seems to be waiting until `CpsThreadGroup` threads are done with their jobs.

Alternatively, we are currently considering if this case couldn't actually be something that happens during a real life Jenkins dirty restart, if Jenkins does shut down in a very specific moment while this thread is still not done, and an expected consequence of it being a "dirty restart".

Reply:
> if this case couldn't actually be something that happens during a real life Jenkins dirty restart

Probably it can.

As I think I mentioned, I am not very familiar with any of the tests in this class, but from a brief look my guess is that `assertIncludesNodes` is just not useful to test and should be deleted unless we can actually make hard guarantees about behavior, which seems unlikely in the absence of transactional storage.

Reply:
If we decide to accept it as acceptable collateral of a dirty restart and do that, what about `testFullyDurableSurvivesDirtyRestart` which also checks the head node (that it is a semaphore step or sleep) in `verifySafelyResumed`? (Going to investigate how the job continues if it loses the sleep node and the head node is set as `{`.)

(Also I wonder if this can't cause a step to run twice - step starts but head is not written in time, shutdown happens, head is set back, step restarts)

Reply:
@jglick I think part of our struggle is in understanding what the expected behavior is for a durable pipeline based on the various states of persistence it can be in at the moment jenkins is killed.  (for example, the question in the previous comment).  Is this something you have some knowledge of (or know someone who does?)

Reply:
> I wonder if this can't cause a step to run twice

I suspect it could, if you were very unlucky. Most builds spend the vast majority of their time inside some `sh` step, during which time there should be no step metadata changing (JENKINS-52165 complicates this a bit—the push mode is cleaner but not yet default). If Jenkins happens to crash at the moment that some metadata is being written, there can be problems. We try to use `AtomicFile` to defend against corruption of a _single_ file (even that is not necessarily reliable, depending on the filesystem; I think #452 etc. pertains to this), but the current simplistic build storage mechanism (which was really intended as an MVP) divides important information between

* `FlowExecution` parts of `build.xml`
* `program.dat`
* `workflow/*.xml`

and these files are not written in a properly journaled order.

> Is this something you have some knowledge of

Not a lot; the author of this test (and the whole system of picking different durability levels) no longer works on Jenkins.

Reply:
If all the head node related assertions were commented out and the pipeline was let continue with that malformed head, it logged "Sleeping for 30 sec" a total of _three_ times (once before restart, twice after) and then a different assertion in verifySucceededCleanly failed because it counted 9 nodes instead of 8. So the pipeline continuing "normally" in this case is quite dubious.

Reply:
Indeed that sounds dubious. If you have the ambition to rewrite the storage system better (and we would have to consider whether to move to a transactional database or something), I am not sure what there is to do except relax these tests, or just delete them if they are not really proving anything interesting.

Reply:
> I wonder if this can't cause a step to run twice

Yes, I have seen steps run twice when trying to create automated tests for edge cases related to resumption and persistence. I feel like I have even seen it in a support ticket or two, but I cannot recall any specifics.

> that assertIncludesNodes is just not useful to test and should be deleted unless we can actually make hard guarantees about behavior

I agree with this. Generally speaking, I think that any test which uses `addStepWithDirtyShutdown` (or `thenWithHardShutdown`) has nondeterministic behavior given the current implementation of `workflow-cps` unless it sets up very precise initial conditions, which IMO kind of defeats the point of testing with `addStepWithDirtyShutdown` in the first place.

Personally, I would first try to swap the tests that use `addStepWithDirtyShutdown` to have only the most basic assertions (e.g., `verifyFailedCleanly`, `verifySafelyResumed`). If even that is not reliable, then I would probably just `@Ignore` them.

Reply:
thanks for the info @dwnusbaum && @jglick , that is very helpful!

Reply:
thank you for the info, it does help!
I'll however go a bit back to `verifySafelyResumed` - this also checks the last node upon restart as well as the final amount of nodes in the finished pipeline. How do we define _safely resumed_ if we skip these assertions? What should we test for here?

Reply:
Since we control the content of the pipeline we should have a good expectation of what a correctly resumed and completed run should look like, and would could just compare the final pipeline state to that.  Unfortunately, I think we end up back in the same place (flake town) because of the unpredictability of `addStepWithDirtyShutdown`.   It seems like `workflow-cps` makes a best effort to safely persist its state so that it can resume but does not make any guarantees.  I am leaning toward more or less ignoring these flakey tests until such time as workflow-cps has been made more bulletproof.

Reply:
> How do we define safely resumed if we skip these assertions?

I think that checking the final build result and that the logs contain some expect content at least once is the most that we can currently expect to be reliable.

I would probably delete the assertions related to `heads` and `node`, replace all calls to `verifySucceededCleanly` with `j.assertBuildStatus(Result.SUCCESS, run); verifyCompletedCleanly(run)`, and then delete `verifySucceededCleanly`.

Reply:
> the unpredictability of `addStepWithDirtyShutdown`

Of course that is the whole purpose of this method—to shake out problems like these.

> `workflow-cps` makes a best effort to safely persist its state

Maybe “some” effort if not the best. :grimacing:

Reply:
maybe the node checks could stay for clean shutdown tests (eg `testDurableAgainstCleanRestartSurvivesIt`), since they do guarantee us correct node persistence AFAIK?

Reply:
Sure, that seems fine assuming we have never seen it flake.